### PR TITLE
Patreon

### LINKED ISSUE #1765: Shoutout our patreon supporters!
In a recent staff meeting we agreed that we weren't giving enough love to our patreon supporters.

As such, we have agreed that we should have the bot post in #community-meta every time someone new receives one of the patreon roles, along with posting once a month (on the 1st, just after the patreon subscription cycle) with a full list of all of our supporters.

The post when we get a new supporter should include:
 - A mention for the user who has started supporting
 - What tier they are supporting us
 - A link to our patreon page

The monthly post should include:
- A 3 column embed (`embed.add_field()` makes this easy), one for each tier
- The names of all of our current supporters, under the appropriate tier column
- A link to our patreon page

### PR DESCRIPTION
**Pull request by mathstrains21**
Closes #1765 

First off, this is my first pull request so please give me general pull request feedback if you have any :-)

Some decisions have been made on discord. A search of `bot#1765` should find you all the conversations.

#### Some changes to make to test:
- Create 3 new roles, with clear different colours so you can see colours on embeds!
- Put these in your config.yml

There are 2 new commands: `!patrons` and `!patreon`. These send the list of patrons and some patreon info respectively.

#### Logging
I have not included an logging. Please provide feedback about logging.

### CODE REVIEW

**Review by TizzySaurus — DISMISSED** [review]
Thanks for the hardwork you've put into this PR!

A few small things in my preliminary review.

**Review by wookie184 — DISMISSED** [review]
I've only tested it quickly so this is mainly a code review (although it seems to work nicely, I like the embeds!)

Firstly, thanks for the PR! The structure of the code seems good and it's been written well.

My main comments here are about repetition, there are couple of places where I think using loops would neaten the code up a bit, making it a bit more concise and extensible. That said, these aren't required changes, i've only mentioned them as something to think about if you want to.

File: bot/exts/info/patreon.py
Comment (Bluenix2) [reply bot/exts/info/patreon.py]: I agree with Wookie here, with a small comment that basically clarifies what that last line does this is more maintainable.

Something like "Remove members who's already in the 'tier_members' list and have a higher role".. maybe?

File: bot/exts/info/patreon.py
Comment (Bluenix2) [reply bot/exts/info/patreon.py:56]: Will this work correctly (both semantically doing what you want to - and not breaking) considering the `enumerate()` is inside the `reversed()`?

File: bot/exts/info/patreon.py
Comment (Bluenix2) [reply bot/exts/info/patreon.py:10]: I think it's still good to keep, since `print()` doesn't work at all if you want to debug something.

File: config-default.yml
Comment (Bluenix2) [inline config-default.yml:297]: Aren't these supposed to be the IDs from the actual Python Discord server?

File: bot/exts/info/patreon.py
Comment (Bluenix2) [inline bot/exts/info/patreon.py:129]: There's no check on this, anybody can send this anywhere.

File: bot/exts/info/patreon.py
Comment (Bluenix2) [inline bot/exts/info/patreon.py:152]:
```suggestion
                "stickers, microcontrollers that support CircuitPython, or maybe even a mechanical keyboard.\n\n"

                "You can" read more about how Patreon supports us, or even support us yourself, on our Patreon page "
```

I like to refect empty lines in the code.

Comment (Bluenix2) [inline bot/exts/info/patreon.py:152]: This is kind of misleading, Patreon the company doesn't support us.

File: bot/exts/info/patreon.py
Comment (Bluenix2) [inline bot/exts/info/patreon.py:146]:
```suggestion
        """Send information about how Python Discord uses Patreon."""
```

File: bot/exts/info/patreon.py
Comment (Bluenix2) [inline bot/exts/info/patreon.py:72]: Yeah I really would like to see some of the suggestions brought up by Wookie implemented.

Even though it may result in more complicated code, that code will take less lines and as such be easier to read (on top of looking cleaner).

File: bot/exts/info/patreon.py
Comment (mathstrains21) [reply bot/exts/info/patreon.py:129]: How would I let admins (or whatever role we want) send it in community meta, while the rest of the community can only send in the bot command channels?

File: config-default.yml
Comment (mathstrains21) [reply config-default.yml:297]: assuming the ones that chris posted in #1765 are correct, then these are correct

File: bot/exts/info/patreon.py
Comment (onerandomusername) [inline bot/exts/info/patreon.py:141]:
```py
if date != 1:
  return
```
This has the benefit of returning early/guarding ifs, which are easier to read.

File: bot/exts/info/patreon.py
Comment (onerandomusername) [inline bot/exts/info/patreon.py:108]: these roles should be an instance attribute.

File: bot/exts/info/patreon.py
Comment (onerandomusername) [inline bot/exts/info/patreon.py:79]: I like using a set here, since it favours no patreon over another, it effectively randomly sorts them.

File: bot/exts/info/patreon.py
Comment (onerandomusername) [inline bot/exts/info/patreon.py:131]: this command name of `patrons` vs `patreon` for the below command, will likely be a source of confusion. A different way to go about this may be making the below informational command a command *group*, and making the list a subcommand. 

The docs for doing so are located here: https://discordpy.readthedocs.io/en/master/ext/commands/api.html#group

File: bot/exts/info/patreon.py
Comment (onerandomusername) [inline bot/exts/info/patreon.py:126]: I was thinking about a way to do this, and think I've come up with a way to do it.

additionally, I think that tier 3 should be before tier 2 and 1, as tier 3 is above tier 2.

Using a dictionary we can match the roles to the members, which ends up like this:
*note: this builds off of storing the patreon roles in a list*
```py
embeds = []
for num, role in reversed(enumerate(patreon_role_list)):
    if not role.members:
       # skipping this role since there are no members
       continue

    # right here the description could be formulated, instead of above

    embeds.append(discord.Embed(
                title=f"Tier {num + 1} patrons",
                description="\n".join(role.members),
                colour=role.colour
            )
    )

if not embeds:
   # there's no embeds, so there's no embed to send
   # yes, this is an extreme edge case to not have any embeds, but should be prepared for it anyhow.

```

**Review by onerandomusername — CHANGES_REQUESTED** [review]
This is really good, honestly, for a first pr, keep it up /gen

File: bot/exts/info/patreon.py
Comment (mathstrains21) [reply bot/exts/info/patreon.py:79]: Most certainly not the intended reason for this set usage :-)

**Comment by mathstrains21** [comment]
> @mathstrains21 will you be continuing this PR?

Hi! I will be, I just need to find the time to do so! Thanks for the reminder though!

**Comment by Xithrius** [comment]
Thanks for the quick update! Have a good one!

**Comment by Xithrius** [comment]
@mathstrains21 Hey, it's been a while. What's your status on this PR?

**Comment by Xithrius** [comment]
@mathstrains21 how's it going?

**Comment by wookie184** [comment]
I'll finish this one off

**Comment by wookie184** [comment]
will have to open a new PR I think

**Comment by mathstrains21** [comment]
I don't have time to continue, but would you like access to my repo to keep it in one PR?

**Comment by wookie184** [comment]
> I don't have time to continue right now, but would you like access to my repo to keep it in one PR?

That would be helpful, thanks!

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py]: Changed, I might add some logging but doesn't hurt to leave it regardless

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py]: Turns out it didn't work, I went for a similar approach that does

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py]: Changed, i'd be happy for any more feedback on the wording though, probably still room for improvement.

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py]: Fixed

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py]: I'm not sure it really makes a difference here, I think i'll leave it as is

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py:105]: I agree it is a bit confusing, i'll think about how best to do this

File: bot/exts/info/patreon.py
Comment (wookie184) [inline bot/exts/info/patreon.py:85]: I'm not sure how well having mentions in embeds will work, given they often don't resolve. Would it be better to make this a normal text message with allowed_mentions to prevent mentioning users. The character limit is much more limiting unless we use 3 separate messages though.

I think I should also add some sort of overflow handling if there are too many users. I'm thinking I might shuffle each tier and pick the first 10 or something.

Comment (wookie184) [reply bot/exts/info/patreon.py:85]: This is how it currently looks on server for me (ignore first embed):

image

Not great. I'll try with a normal text message

Comment (minalike) [reply bot/exts/info/patreon.py:85]: Since you're likely going to remove the mentions, why not list all the users in paragraph form? Then we shouldn't need to pick only a random 10.

Comment (ionite34) [reply bot/exts/info/patreon.py:85]: Should we consider maybe multi-column tables or embeds? Should at least be able to fit in all the patron supporters that way.

img1

img2

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py]: I have limited the command to list patrons to `bot_commands`

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py]: Done

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py]: I've gone with plain text names listed in the embeds, I think that's neatest.

**Comment by wookie184** [comment]
This is now ready for review

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py]: I have refactored this.

**Comment by onerandomusername** [comment]
Please remove my request for review.

File: bot/exts/info/patreon.py
Comment (onerandomusername) [inline bot/exts/info/patreon.py:80]:
```suggestion
            patron_names = [f"• {str(patron.name)}" for patron in patrons]
```

**Review by ionite34 — DISMISSED** [review]

File: bot/exts/info/patreon.py
Comment (onerandomusername) [inline bot/exts/info/patreon.py:58]: This has an inherent dependency on how the patreon bot works. If the patreon bot decided to take away and giveth the new role in two seperate api requests, this would cease to work. Or, if the patreon bot were to take the old role away and then give the new role, this would also cease to work.

This all hinges on the patreon bot making a single api request, and would lead to funky behaviour if their bot used more than one request--and I believe it does.

**Review by onerandomusername — COMMENTED** [review]
didn't intend to leave another review but I am.

File: bot/exts/info/patreon.py
Comment (TizzySaurus) [reply bot/exts/info/patreon.py:80]: I believe you meant `str(patron)`, which since this is an f-string would be the same as just `patron`.

File: bot/exts/info/patreon.py
Comment (TizzySaurus) [inline bot/exts/info/patreon.py:18]: Minor, but "Python Discord is a volunteer led non-profit, so we rely..." sounds slightly better to me

File: bot/exts/info/patreon.py
Comment (TizzySaurus) [inline bot/exts/info/patreon.py:38]: Again minor, but this should be "A patreon tier of 0 indicates the user is not a *patron*."

File: bot/exts/info/patreon.py
Comment (TizzySaurus) [inline bot/exts/info/patreon.py:60]: I think `if new_patreon_tier <= old_patreon_tier` reads nicer here.

File: bot/exts/info/patreon.py
Comment (TizzySaurus) [inline bot/exts/info/patreon.py:79]: Since `PATREON_TIERS` is ordered highest to lowest, that means we iterate highest to lowest.

As such, we could instead store "seen patrons" and check if they're in that list. May be slightly cleaner.

File: bot/exts/info/patreon.py
Comment (TizzySaurus) [inline bot/exts/info/patreon.py:51]: I believe we may need to add `await self.bot.wait_until_ready()` here to ensure that the cache will be loaded before running tasks (ensure L72 + L76 etc. will definitely work).

File: bot/exts/info/patreon.py
Comment (TizzySaurus) [inline bot/exts/info/patreon.py:91]: I don't think it's necessary for this to be split over multiple lines, and IMO one line is cleaner.

File: bot/exts/info/patreon.py
Comment (TizzySaurus) [inline bot/exts/info/patreon.py:97]: According to [the docs](https://discordpy.readthedocs.io/en/latest/api.html#discord.abc.Messageable.send), the embeds kwarg should be a list and not a tuple.

File: bot/exts/info/patreon.py
Comment (TizzySaurus) [inline bot/exts/info/patreon.py:100]: Do we want to also limit what channels this can be invoked in? It's a much smaller message than the supporters subcommand, but the default attitude seems to be to limit things to bot-commands (maybe also pygen and/or community-meta here).

**Review by TizzySaurus — COMMENTED** [review]
A few points, mostly about style so not requesting changes so as to not block the PR with this review.

File: bot/exts/info/patreon.py
Comment (onerandomusername) [reply bot/exts/info/patreon.py:80]: Yep, edited the suggestion

Comment (TizzySaurus) [reply bot/exts/info/patreon.py:80]: The `str()` part isn't necessary (f-strings automatically call the `__str__()`)

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py:51]: We wait until ready before loading extensions so it shouldn't matter https://github.com/python-discord/bot-core/blob/127d0efa219d29e15bd71339d0d2517945aed57c/botcore/_bot.py#L125

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py:91]: It would make a pretty long line. I prefer it as is.

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py:79]: This already feels pretty clean and explicit to me, keeping track of an extra list would require more code.

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py:100]: Given this is static text it feels similar to a tag which we allow to be invoked anywhere. I don't feel strongly either way though.

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py:58]: I don't think this should be an issue. Whatever order things happen there will still only be one member_update that involves the tier increasing, which is all that this checks.

The only slightly unintended case would be if a user moves down a tier (e.g. from 3 to 2), and the bot does this by removing 3 and then adding 2. In this case the message would be sent, although it still wouldn't be incorrect (Just saying they are now a tier 2 patron).

Fixing this would probably require waiting for other role changes to see if they change things, although this could be fairly complicated so I think it would be best to see how what we have currently works out.

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py:97]: I opened https://github.com/Rapptz/discord.py/issues/8460 as this shouldn't matter and it's type hinted as allowing sequences anyway

File: bot/exts/info/patreon.py
Comment (TizzySaurus) [reply bot/exts/info/patreon.py:79]: Fair enough :+1:

File: bot/exts/info/patreon.py
Comment (TizzySaurus) [reply bot/exts/info/patreon.py:91]: It would be less than the 120 char limit (107 chars), but ultimately up to you.

File: bot/exts/info/patreon.py
Comment (TizzySaurus) [reply bot/exts/info/patreon.py:99]: I don't feel strongly either, and I don't see it being an issue so think it's fine as-is. If it does prove to be an issue we can always add in another PR.

File: bot/exts/info/patreon.py
Comment (TizzySaurus) [inline bot/exts/info/patreon.py:17]: I think you accidentally removed the "is" after Python Discord

File: bot/exts/info/patreon.py
Comment (TizzySaurus) [inline bot/exts/info/patreon.py:19]: NIT: "and maybe" --> "and sometimes"

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py]: 👍 I made a couple other small changes to it too

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py:17]: oops

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py]: I don't think we've ever actually given out a mechanical keyboard, so I removed it from here.

**Review by ionite34 — APPROVED** [review]
Wasn't able to test the role assignments on real users, but seem to work fine in manual testing.

File: bot/exts/info/patreon.py
Comment (ChrisLovering) [inline bot/exts/info/patreon.py:79]: If this list is empty, imo the embed should contain something like "No patrons in this tier" as opposed to being something like this:
image

**Review by ChrisLovering — APPROVED** [review]
Approving as my only comment won't affect us atm, so I'll leave it to you to decide whether to action it before merging.

File: bot/exts/info/patreon.py
Comment (wookie184) [reply bot/exts/info/patreon.py:80]: That makes sense, done 👍

**Comment by mathstrains21** [comment]
Thanks @wookie184 for finishing this off for me!

**Pull request merged**
